### ISSUE TITLE
The fifth formula in the paper

### ISSUE THREAD

**Issue by cervineee** [LanglandsLin/ActCLR#17]
Hello, the fifth formula in the paper shows that the channel dimensions are weighted and summed, but it is not reflected in the code but is averaged. Can you tell me why this is so?         cam = F.relu(cam.mean(dim=1, keepdim=True)) / (1e-8)

**Comment by LanglandsLin** [comment]
https://github.com/LanglandsLin/ActCLR/blob/9b1b4be17f33bf2479f72c7987588c3c909325b1/net/st_gcn.py#L94

Here the weights are calculated and multiplied to the channel.

**Comment by cervineee** [comment]
I'm very sorry to bother you again. Why is the mean operation performed here?cam = F.relu(cam.mean(dim=1, keepdim=True)) / (1e-8)

**Comment by cervineee** [comment]
Sorry to bother you, why do you use the index node list in the -1 dimension instead of using each node of the adjacency matrix?
            for body_part in body_parts:
                cam[:, :, :, body_part] = cam[:, :, :, body_part].mean(dim=-1, keepdim=True)
This step of loop processing does not affect the original tensor cam. Why do you do this?

**Comment by LanglandsLin** [comment]
> I'm very sorry to bother you again. Why is the mean operation performed here?cam = F.relu(cam.mean(dim=1, keepdim=True)) / (1e-8)

This is a common operation in grad cam.
[Grad-CAM: Visual Explanations from Deep Networks via Gradient-based Localization](https://arxiv.org/pdf/1610.02391.pdf)

**Comment by LanglandsLin** [comment]
> Sorry to bother you, why do you use the index node list in the -1 dimension instead of using each node of the adjacency matrix? for body_part in body_parts: cam[:, :, :, body_part] = cam[:, :, :, body_part].mean(dim=-1, keepdim=True) This step of loop processing does not affect the original tensor cam. Why do you do this?

We note that actions usually take place in parts, so we do smoothing inside the parts to make them have the same activation weight.

**Comment by cervineee** [comment]
> > 很抱歉再次打扰您。为什么在这里执行平均运算？cam = F.relu（cam.mean（dim=1， keepdim=True）） / （1e-8）
> 
> 这是grad cam中的常见操作。 [Grad-CAM：通过基于梯度的定位实现深度网络的视觉解释](https://arxiv.org/pdf/1610.02391.pdf)
Thank you for your answer, I went to check the relevant paper description, L
c
After Grad-CAM (Formula 2), there is no specific information about performing the mean operation, including in your paper, in Formula 5, the relu information is cam = dz_dy_mean[:,:,None,None] * y, which is not specific. Mention the reasons for performing mean operations on channel dimensions


> > I'm very sorry to bother you again. Why is the mean operation performed here?cam = F.relu(cam.mean(dim=1, keepdim=True)) / (1e-8)
> 
> This is a common operation in grad cam. [Grad-CAM: Visual Explanations from Deep Networks via Gradient-based Localization](https://arxiv.org/pdf/1610.02391.pdf)
Thank you for your answer, I went to check the relevant paper description, LcGrad-CAM (Formula 2), there is no specific information about performing the mean operation, including in your paper, in Formula 5, the relu information is cam = dz_dy_mean[:,:,None,None] * y, which is not specific mention the reasons for performing mean operations on channel dimensions

**Comment by cervineee** [comment]
> > I'm very sorry to bother you again. Why is the mean operation performed here?cam = F.relu(cam.mean(dim=1, keepdim=True)) / (1e-8)
> 
> This is a common operation in grad cam. [Grad-CAM: Visual Explanations from Deep Networks via Gradient-based Localization](https://arxiv.org/pdf/1610.02391.pdf)

Thank you for your answer, I went to check the relevant paper description, LcGrad-CAM (Formula 2), there is no specific information about performing the mean operation, including in your paper, in Formula 5, the relu information is cam = dz_dy_mean[:,:,None,None] * y, which is not specific mention the reasons for performing mean operations on channel dimensions

**Comment by LanglandsLin** [comment]
cam equals $\alpha$, and y is $A$ in Grad-CAM. The paper Grad-CAM uses summation, we use mean. It's the same because we divide by a factor.

**Comment by cervineee** [comment]
> https://github.com/LanglandsLin/ActCLR/blob/9b1b4be17f33bf2479f72c7987588c3c909325b1/net/st_gcn.py#L94
> 
> Here the weights are calculated and multiplied to the channel.

Thank you very much for your answer. Due to my own reasons, I still have doubts about the mean operation. I think that in the calculation expressed in Formula 5 in the paper, a weighted sum cam = dz_dy_mean[:,:,None,None has been performed ] * y
The subsequent cam.mean(dim=1, keepdim=True) is not reflected in the formula and is a redundant calculation, so I have been asking you about the role of mean calculation.

**Comment by LanglandsLin** [comment]
``cam = dz_dy_mean[:,:,None,None] * y" only performs channel-wise weighting, and ``cam.mean(dim=1, keepdim=True)"  performs channel-wise summation.

**Comment by LanglandsLin** [comment]
发自我的 iPhone在 2024年7月19日，16:14，cervineee ***@***.***> 写道：﻿

https://github.com/LanglandsLin/ActCLR/blob/9b1b4be17f33bf2479f72c7987588c3c909325b1/net/st_gcn.py#L94
Here the weights are calculated and multiplied to the channel.

Thank you very much for your answer. Due to my own reasons, I still have doubts about the mean operation. I think that in the calculation expressed in Formula 5 in the paper, a weighted sum cam = dz_dy_mean[:,:,None,None has been performed ] * y
The subsequent cam.mean(dim=1, keepdim=True) is not reflected in the formula and is a redundant calculation, so I have been asking you about the role of mean calculation.
